src/lazy.jl

```diff
@@ -0,0 +1,722 @@
+            pos += 1
+            pos > len && return pos
+            b = getbyte(buf, pos)
+        elseif !foundr
+            # if we didn't find a newline and we're not EOF
+            # then that's an error; only whitespace, newlines,
+            # and EOFs are valid in between lines
+            error = ExpectedNewline
+            @goto invalid
+        end
+        # since we found a newline, we now ignore all whitespace, including newlines (empty lines)
+        # until we find EOF or non-whitespace
+        while b == UInt8(' ') || b == UInt8('\t') || b == UInt8('\n') || b == UInt8('\r')
+            pos += 1
+            pos > len && return pos
+            b = getbyte(buf, pos)
+        end
+    end)
+end
+
+# core JSON array parsing function
+# takes a `keyvalfunc` that is applied to each index => value element
+# `keyvalfunc` is provided a Int => LazyValue pair
+# applyeach always requires a key-value pair function
+# so we use the index as the key
+# returns a `pos` value that notes the next position where parsing should continue
+# this is essentially the `StructUtils.applyeach` implementation for LazyValues w/ type ARRAY
+function applyarray(keyvalfunc, x::LazyValues)
+    pos = getpos(x)
+    buf = getbuf(x)
+    len = getlength(buf)
+    opts = getopts(x)
+    jsonlines = opts.jsonlines
+    b = getbyte(buf, pos)
+    if !jsonlines
+        if b != UInt8('[')
+            error = ExpectedOpeningArrayChar
+            @goto invalid
+        end
+        pos += 1
+        @nextbyte
+        b == UInt8(']') && return pos + 1
+    else
+        # for jsonlines, we need to make sure that recursive
+        # lazy values *don't* consider individual lines *also*
+        # to be jsonlines
+        opts = LazyOptions(; allownan=opts.allownan, ninf=opts.ninf, inf=opts.inf, nan=opts.nan, jsonlines=false)
+    end
+    i = 1
+    while true
+        # we're now positioned at the start of the value
+        val = lazy(buf, pos, len, b, opts)
+        ret = keyvalfunc(i, val)
+        ret isa StructUtils.EarlyReturn && return ret
+        # if keyvalfunc didn't materialize `val` and return an
+        # updated `pos`, then we need to skip val ourselves
+        # WARNING: parsing can get corrupted if random Int values are returned from keyvalfunc
+        pos = (ret isa Int && ret > pos) ? ret : skip(val)
+        if jsonlines
+            @jsonlines_checks
+        else
+            @nextbyte
+            if b == UInt8(']')
+                return pos + 1
+            elseif b != UInt8(',')
+                error = ExpectedComma
+                @goto invalid
+            end
+            pos += 1 # move past ','
+            @nextbyte
+        end
+        i += 1
+    end
+
+@label invalid
+    invalid(error, buf, pos, "array")
+end
+
+# temporary string type to enable deferrment of string allocation in certain cases (like navigating a lazy structure)
+struct PtrString
+    ptr::Ptr{UInt8}
+    len::Int
+    escaped::Bool
+end
```

Comment:
Why not use a view instead of a ptr and len here?

Reply:
Performance. It's much slower to use a view in terms of allocating the struct vs. the minimal Ptr + len for the very-short-lived time these are alive. I tried 3-4 different variations of possible ways to handle these and this was the most performant along with the guards to ensure PtrString is never visible at the user-level.

Reply:
I would be curious if a `MemoryRef` based approach could have equal/better perf here.